src/kernel/boot.c

```diff
@@ -838,34 +837,88 @@ BOOT_CODE bool_t init_freemem(word_t n_available, const p_region_t *available,
-                ndks_boot.freemem[next].start = start + size;
+        region_t *reg = &ndks_boot.freemem[i];
+        if (is_reg_empty(*reg)) {
+            /* Skip any empty regions. From the check above we know there must
+             * be at least one at the end of the list. We do not really expect
+             * to see empty regions in the middle of the array.
+             */
+            continue;
+        }
+        if (reg->end - reg->start >= size) {
+            pptr_t start = ROUND_DOWN(reg->end - size, max_bits);
+            if (start >= reg->start) {
+                /* Found a region for the root server objects. */
+                rootserver_mem.start = start;
+                rootserver_mem.end = start + size;
+                break;
             }
-            ndks_boot.freemem[i].end = start;
-            break;
-        } else if (i < ARRAY_SIZE(ndks_boot.freemem)) {
-            ndks_boot.freemem[next] = ndks_boot.freemem[i];
         }
+        /* The region is too small. Since we loop though the list from the end
+         * and we know there is an empty region at the end, we can be sure that
+         * if we arrive here, there is an empty region in the next higher slot.
+         * Swap the slots and try the next lower region:
+         *
+         *                region too small
+         *                v
+         * Free: [a] [b] [c] [empty] [...]  ->  [a] [b] [empty] [c] [...]
+         *            ^
+         *            region to check next
+         */
+        unsigned int i_next = i + 1;
+        assert(i_next < ARRAY_SIZE(ndks_boot.freemem));
+        region_t *next = &ndks_boot.freemem[i_next];
+        assert(is_reg_empty(*next));
+        *next = *reg;
+        *reg = REG_EMPTY;
+    }
+
+    if (i < 0) {
+        printf("ERROR: no free memory region is big enough for root server "
+               "objects, need size/alignment of 2^%"SEL4_PRIu_word"\n",
+               max_bits);
+        /* Fatal error, can't create root server objects. */
+        return false;
     }
 
+    /* Carve out the memory for the region of the rootserver objects from the
+     * free memory list and put the regions before and after it in instead:
+     *
+     * Free: [a] [b] [c] [empty] [...] -> [a] [b] [pre] [post] [...]
+     *                ^
+     *                region for rootserver objects
+     */
+    region_t *reg_pre = &ndks_boot.freemem[i];
+    region_t *reg_post = &ndks_boot.freemem[i + 1];
+    assert(is_reg_empty(*reg_post));
+    *reg_post = (region_t) { /* fill the empty slot */
+        .start = rootserver_mem.end,
+        .end = reg_pre->end
+    };
+    reg_pre->end = rootserver_mem.start; /* shrink the region */
+
+    /* The free memory list has been updated, create the root server objects
+     * from the memory block carved out and now described in 'rootserver_mem'.
+     */
+    create_rootserver_objects(it_v_reg, extra_bi_size_bits);
```

Comment:
This doesn't look like a 'fix' to me but a rewrite. Will this change the ordering of memory objects? 

@lsf37 I vaguely recall the trusted loader / capdl / camkes relied on python code that emulated how the boot code assigns regions in order to do offline allocation. 

Does this change that?

IMO still wrong direction - we should be fixing minimal bugs, keeping the current layout, and revisiting the boot code as a whole with a verification plan and the ability to do offline allocaiton, as well as supporting other requirements (like swapping out boot code).

Reply:
@corlewis probably knows the current capDL loader assumptions best. @corlewis can you comment on that?

Reply:
This does not change anything in the algorithm itself, it's just more verbose now. It also moves things out of the loop that just tries to find a place.  After the loop it catches the error now if no suitable region was found - which was missing before, but usually worked because the regions are big enough.

Reply:
> I vaguely recall the trusted loader / capdl / camkes relied on python code that emulated how the boot code assigns regions in order to do offline allocation.

That's my understanding as well. See https://github.com/seL4/capdl/commit/ef2fe74242214da796179cab15821b66e7468a07 for what I think is the original commit dealing with this.

Reply:
>  After the loop it catches the error now if no suitable region was found - which was missing before, but usually worked because the regions are big enough.

Adding this check only requires making the loop variable signed right?  It doesn't require moving the `create_rootserver_objects()` outside of the loop.

Reply:
Making the variable unsigned is essential for the loop to terminate. An unsigned values would just roll over and `i >= 0` is always true. 
The update of the free regions and the call to `create_rootserver_objects()` can be left inside the loop also. Then we could return 'true' there instead of doing a `break`. If we do this, we could unconditionally return `false`when the loop terminates and don't have to check explicitly `i < 0` to catch the error.
My intention of moving things out of the loop was keeping it tighter and just use it for searching a region. Then the function has a way to continue on success in case one wants to add more code here and exists early on any error.